File: Plugin/ShadowWarden/src/main/java/dev/didelfo/shadowWarden/security/E2EE/EphemeralKeyStore.java
package dev.didelfo.shadowWarden.security.E2EE;

import javax.crypto.*;
import javax.crypto.spec.*;
import java.security.*;
import java.security.spec.ECGenParameterSpec;
import java.util.Arrays;

public class EphemeralKeyStore {
    private KeyPair keyPair;

    // Generar par de claves ECDH
    public void generateKeyPair() {
        KeyPairGenerator keyPairGenerator = null;
        try {
            keyPairGenerator = KeyPairGenerator.getInstance("EC");
            ECGenParameterSpec ecGenParameterSpec = new ECGenParameterSpec("secp521r1");
            keyPairGenerator.initialize(ecGenParameterSpec, new SecureRandom());
            keyPair = keyPairGenerator.generateKeyPair();
        } catch (NoSuchAlgorithmException | InvalidAlgorithmParameterException e) {
            throw new RuntimeException(e);
        }

    }

    // Obtener clave pública
    public PublicKey getPublicKey() {
        return keyPair.getPublic();
    }

    // Generar clave compartida
    public byte[] getSharedSecret(PublicKey peerPublicKey) {
        KeyAgreement keyAgreement = null;
        try {
            keyAgreement = KeyAgreement.getInstance("ECDH");
            keyAgreement.init(keyPair.getPrivate());
            keyAgreement.doPhase(peerPublicKey, true);

            // Deriva la clave compartida
            return deriveKey(keyAgreement.generateSecret());

        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
            throw new RuntimeException(e);
        }
    }

    // Deriva clave AES
    private byte[] deriveKey(byte[] sharedSecret) throws NoSuchAlgorithmException {
        MessageDigest messageDigest = MessageDigest.getInstance("SHA-512");
        byte[] hash = messageDigest.digest(sharedSecret);
        return Arrays.copyOf(hash, 32); // 256 bits para AES
    }

    // Genera clave HMAC
    public SecretKey getHmacKey(byte[] sharedSecret) {
        MessageDigest messageDigest = null;
        try {
            messageDigest = MessageDigest.getInstance("SHA-512");
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
        byte[] hmacKeyBytes = messageDigest.digest(
                concatenate(sharedSecret, "HMAC".getBytes()));
        hmacKeyBytes = Arrays.copyOf(hmacKeyBytes, 64); // 512 bits para HMAC-SHA512
        return new SecretKeySpec(hmacKeyBytes, "HmacSHA512");
    }

    // Cifrar y firmar
    public Pair<byte[], byte[]> encryptAndSign(String message, byte[] sharedSecret, SecretKey hmacKey) {

        try {
            // Cifrado AES-256-GCM
            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
            byte[] iv = new byte[12];
            new SecureRandom().nextBytes(iv);
            SecretKeySpec keySpec = new SecretKeySpec(sharedSecret, "AES");
            GCMParameterSpec parameterSpec = new GCMParameterSpec(128, iv);

            cipher.init(Cipher.ENCRYPT_MODE, keySpec, parameterSpec);
            byte[] cipherText = cipher.doFinal(message.getBytes());

            // IV + cipherText
            byte[] encryptedData = concatenate(iv, cipherText);

            // Firmar con HMAC
            Mac mac = Mac.getInstance("HmacSHA512");
            mac.init(hmacKey);
            byte[] signature = mac.doFinal(encryptedData);

            return new Pair<>(encryptedData, signature);
        } catch (NoSuchPaddingException | IllegalBlockSizeException | NoSuchAlgorithmException |
                 InvalidAlgorithmParameterException | BadPaddingException | InvalidKeyException e) {
            return null;
        }
    }

    // Verificar y descifrar
    public String verifyAndDecrypt(byte[] encryptedData, byte[] signature, byte[] sharedSecret, SecretKey hmacKey) {

        try {
            // Verificar firma
            Mac mac = Mac.getInstance("HmacSHA512");
            mac.init(hmacKey);
            byte[] calculatedSignature = mac.doFinal(encryptedData);

            if (!MessageDigest.isEqual(calculatedSignature, signature)) {
                throw new SecurityException("HMAC verification failed");
            }

            // Descifrar
            byte[] iv = Arrays.copyOfRange(encryptedData, 0, 12);
            byte[] cipherText = Arrays.copyOfRange(encryptedData, 12, encryptedData.length);

            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
            SecretKeySpec keySpec = new SecretKeySpec(sharedSecret, "AES");
            GCMParameterSpec parameterSpec = new GCMParameterSpec(128, iv);

            cipher.init(Cipher.DECRYPT_MODE, keySpec, parameterSpec);
            byte[] decryptedBytes = cipher.doFinal(cipherText);

            return new String(decryptedBytes);
        } catch (NoSuchPaddingException | IllegalBlockSizeException | NoSuchAlgorithmException |
                 InvalidAlgorithmParameterException | BadPaddingException | InvalidKeyException e) {
            throw new RuntimeException(e);
        }
    }


    // Método auxiliar para concatenar arrays
    private byte[] concatenate(byte[] a, byte[] b) {
        byte[] result = new byte[a.length + b.length];
        System.arraycopy(a, 0, result, 0, a.length);
        System.arraycopy(b, 0, result, a.length, b.length);
        return result;
    }

    // Clase Pair simple para Java
    public static class Pair<F, S> {
        public final F first;
        public final S second;

        public Pair(F first, S second) {
            this.first = first;
            this.second = second;
        }
    }
}

File: APP/ShadowWarden/app/src/main/java/dev/didelfo/shadowwarden/security/E2EE/EphemeralKeyStore.kt
package dev.didelfo.shadowwarden.security.E2EE


import java.security.*
import java.security.spec.ECGenParameterSpec
import javax.crypto.*
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import java.util.*

object EphemeralKeyStore {
    // variables
    private var keyPair: KeyPair? = null
    private var sharedSecret: ByteArray? = null
    private var hmacKey: SecretKey? = null

    init {
        // Inicializamos el singelton
        generateKeyPair()
    }

    // Generar par de claves ECDH
    fun generateKeyPair() {
        clearKeys() // Limpiar claves anteriores
        val keyPairGenerator = KeyPairGenerator.getInstance("EC")
        val ecGenParameterSpec = ECGenParameterSpec("secp521r1")
        keyPairGenerator.initialize(ecGenParameterSpec, SecureRandom())
        keyPair = keyPairGenerator.generateKeyPair()
    }

    // Obtener clave pública para compartir
    fun getPublicKey(): PublicKey? {
        return keyPair?.public
    }

    // Generar clave compartida a partir de la clave pública del otro extremo
    fun generateSharedSecret(peerPublicKey: PublicKey) {
        val keyAgreement = KeyAgreement.getInstance("ECDH")
        keyAgreement.init(keyPair?.private)
        keyAgreement.doPhase(peerPublicKey, true)

        // Deriva la clave compartida usando KDF
        sharedSecret = deriveKey(keyAgreement.generateSecret())

        // Generar clave HMAC a partir del secreto compartido
        hmacKey = generateHmacKey(sharedSecret!!)
    }

    // Deriva una clave adecuada para AES-256-GCM
    private fun deriveKey(sharedSecret: ByteArray): ByteArray {
        val messageDigest = MessageDigest.getInstance("SHA-512")
        return messageDigest.digest(sharedSecret).copyOf(32) // Tomamos 256 bits (32 bytes)
    }

    // Genera clave HMAC a partir del secreto compartido
    private fun generateHmacKey(sharedSecret: ByteArray): SecretKey {
        val messageDigest = MessageDigest.getInstance("SHA-512")
        val hmacKeyBytes = messageDigest.digest(sharedSecret + "HMAC".toByteArray()).copyOf(64)
        return SecretKeySpec(hmacKeyBytes, "HmacSHA512")
    }

    // Cifrar mensaje con AES-256-GCM y firmar con HMAC-SHA512
    fun encryptAndSign(message: String): Pair<ByteArray, ByteArray> {
        if (sharedSecret == null) throw IllegalStateException("Shared secret not generated")

        // Cifrado AES-256-GCM
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val iv = ByteArray(12).also { SecureRandom().nextBytes(it) }
        val keySpec = SecretKeySpec(sharedSecret!!, "AES")
        val parameterSpec = GCMParameterSpec(128, iv)

        cipher.init(Cipher.ENCRYPT_MODE, keySpec, parameterSpec)
        val cipherText = cipher.doFinal(message.toByteArray(Charsets.UTF_8))

        // Concatenar IV + cipherText
        val encryptedData = iv + cipherText

        // Firmar con HMAC-SHA512
        val mac = Mac.getInstance("HmacSHA512")
        mac.init(hmacKey)
        val signature = mac.doFinal(encryptedData)

        return Pair(encryptedData, signature)
    }

    // Verificar y descifrar mensaje
    fun verifyAndDecrypt(encryptedData: ByteArray, signature: ByteArray): String {
        if (sharedSecret == null) throw IllegalStateException("Shared secret not generated")
        if (hmacKey == null) throw IllegalStateException("HMAC key not generated")

        // Verificar firma HMAC
        val mac = Mac.getInstance("HmacSHA512")
        mac.init(hmacKey)
        val calculatedSignature = mac.doFinal(encryptedData)

        if (!MessageDigest.isEqual(calculatedSignature, signature)) {
            throw SecurityException("HMAC verification failed")
        }

        // Descifrar AES-256-GCM
        val iv = encryptedData.copyOfRange(0, 12)
        val cipherText = encryptedData.copyOfRange(12, encryptedData.size)

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val keySpec = SecretKeySpec(sharedSecret!!, "AES")
        val parameterSpec = GCMParameterSpec(128, iv)

        cipher.init(Cipher.DECRYPT_MODE, keySpec, parameterSpec)
        val decryptedBytes = cipher.doFinal(cipherText)

        return String(decryptedBytes, Charsets.UTF_8)
    }

    // Limpiar claves de memoria
    fun clearKeys() {
        keyPair = null
        sharedSecret?.fill(0)
        sharedSecret = null
        hmacKey = null
    }
}